{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "def load_data(path):\n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data):\n",
    "    training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "    x_train = []\n",
    "    \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "        \n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "            num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((len(num_seq),4))\n",
    "\n",
    "        for i in range (len(num_seq)):\n",
    "            if num_seq[i] <= 3:\n",
    "                X[i,num_seq[i]] = 1               \n",
    "               \n",
    "        x_train.append(X.flatten())\n",
    "\n",
    "\n",
    "    x_train = np.array(x_train)\n",
    "    x_train = np.expand_dims(x_train, axis=2)\n",
    "    return x_train\n",
    "        \n",
    "  \n",
    "def get_class(protein, t_data):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"data/clip/%s/5000/test_sample_0\"% protein)\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "protein = \"1_PARCLIP_AGO1234_hg19\"\n",
    "with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "  cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "\n",
    "    \n",
    "#print (cobinding_data.shape)\n",
    "\n",
    "cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "#print (cobinding.shape)\n",
    "\n",
    "for n in range(0,cobinding_data.shape[1],101):\n",
    "    a = cobinding_data[:,n:(n+101)]\n",
    "    a[0,0] = 1 #zgolj da preverim kam se zapiše a ODSTRANI!\n",
    "    cobinding[:,:,(n/101)] = a\n",
    "    #print (a, a.shape)\n",
    "   \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "#print (cobinding)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein = \"1_PARCLIP_AGO1234_hg19\"\n",
    "with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "  cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "\n",
    "\n",
    "cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "for n in range(0,cobinding_data.shape[1],101):\n",
    "    a = cobinding_data[:,n:(n+101)]\n",
    " \n",
    "    cobinding[:,:,(n/101)] = a\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_25 (Conv1D)           (None, 54, 120)           149880    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_13 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_26 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_11  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_21 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_22 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 155,774\n",
      "Trainable params: 155,774\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_27 (Conv1D)           (None, 54, 120)           149880    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_14 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_28 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_12  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_23 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_24 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 155,774\n",
      "Trainable params: 155,774\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_29 (Conv1D)           (None, 54, 120)           149880    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_15 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_30 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_13  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_25 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_26 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 155,774\n",
      "Trainable params: 155,774\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "(array([[0.9206887 , 0.03235392],\n",
      "       [0.9206887 , 0.03235392],\n",
      "       [0.9242923 , 0.03048423],\n",
      "       ...,\n",
      "       [0.9206887 , 0.03235392],\n",
      "       [0.9206887 , 0.03235392],\n",
      "       [0.9206887 , 0.03235392]], dtype=float32), 0.8304299230896918, (1666, 2))\n"
     ]
    }
   ],
   "source": [
    "for x in range (0,1):\n",
    "   \n",
    "    X = get_seq(protein,\"train\")\n",
    "    y = get_class(protein,\"train\")\n",
    "        \n",
    "    kf = KFold(n_splits=3) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 48,data_format='channels_last', input_shape=(101,26), strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(12, 2, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "        print(model.summary())  \n",
    "        \n",
    "        cobinding_train, cobinding_test = cobinding[train_index], cobinding[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(cobinding_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "    y_scores =(model.predict(cobinding_test))# pridobim napovedi za sekvence iz cobinding\n",
    "    score = roc_auc_score(y_test, y_scores)      \n",
    "    print(y_scores, score, y_scores.shape)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5000, 404, 1)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_6 (Conv1D)            (None, 120, 23)           48600     \n",
      "_________________________________________________________________\n",
      "max_pooling1d_3 (MaxPooling1 (None, 55, 23)            0         \n",
      "_________________________________________________________________\n",
      "conv1d_7 (Conv1D)            (None, 54, 12)            564       \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_3 ( (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_5 (Dense)              (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_6 (Dense)              (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 52,166\n",
      "Trainable params: 52,166\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_8 (Conv1D)            (None, 120, 23)           48600     \n",
      "_________________________________________________________________\n",
      "max_pooling1d_4 (MaxPooling1 (None, 55, 23)            0         \n",
      "_________________________________________________________________\n",
      "conv1d_9 (Conv1D)            (None, 54, 12)            564       \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_4 ( (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_7 (Dense)              (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_8 (Dense)              (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 52,166\n",
      "Trainable params: 52,166\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_10 (Conv1D)           (None, 120, 23)           48600     \n",
      "_________________________________________________________________\n",
      "max_pooling1d_5 (MaxPooling1 (None, 55, 23)            0         \n",
      "_________________________________________________________________\n",
      "conv1d_11 (Conv1D)           (None, 54, 12)            564       \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_5 ( (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_9 (Dense)              (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_10 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 52,166\n",
      "Trainable params: 52,166\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "(array([[0.8421817 , 0.05297954],\n",
      "       [0.8421817 , 0.05297954],\n",
      "       [0.67087793, 0.21986245],\n",
      "       ...,\n",
      "       [0.8421817 , 0.05297954],\n",
      "       [0.8421817 , 0.05297953],\n",
      "       [0.8421817 , 0.05297953]], dtype=float32), 0.8404979804127706, (1666, 2))\n"
     ]
    }
   ],
   "source": [
    "for x in range (0,1):\n",
    "   \n",
    "    X = get_seq(protein,\"train\")\n",
    "    y = get_class(protein,\"train\")\n",
    "        \n",
    "    kf = KFold(n_splits=3) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 4,data_format='channels_first', input_shape=(101,26), strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(12, 2, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "        print(model.summary())  \n",
    "        \n",
    "        cobinding_train, cobinding_test = cobinding[train_index], cobinding[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(cobinding_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "    y_scores =(model.predict(cobinding_test))# pridobim napovedi za sekvence iz cobinding\n",
    "    score = roc_auc_score(y_test, y_scores)      \n",
    "    print(y_scores, score, y_scores.shape)\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#spreminjam obliko podatkov za sekvenco; spremenjena koda za get_seq\n",
    "\n",
    "training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "x_train = []\n",
    "    \n",
    "    \n",
    "for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "        \n",
    "        for i in range(0,len(sequence)):\n",
    "            num_seq.append(nucleotide[sequence[i]])\n",
    "            \n",
    "        \n",
    "        x_train.append(num_seq)\n",
    "\n",
    "x_train = np.array(x_train)\n",
    "\n",
    "        \n",
    "\n",
    "#x_train.append(X.flatten())\n",
    "\n",
    "#x_train = np.expand_dims(x_train, axis=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5000, 101)\n"
     ]
    }
   ],
   "source": [
    "print(x_train.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5000, 101, 26)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(5000, 101, 27)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#uporabim na novo oblikovano sekvenco\n",
    "\n",
    "print(cobinding.shape)\n",
    "X = np.dstack((cobinding, x_train))\n",
    "X.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_32 (Conv1D)           (None, 54, 120)           155640    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_16 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_33 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_16  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_31 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_32 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 161,534\n",
      "Trainable params: 161,534\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_34 (Conv1D)           (None, 54, 120)           155640    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_17 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_35 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_17  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_33 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_34 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 161,534\n",
      "Trainable params: 161,534\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_36 (Conv1D)           (None, 54, 120)           155640    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_18 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_37 (Conv1D)           (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_18  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_35 (Dense)             (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_36 (Dense)             (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 161,534\n",
      "Trainable params: 161,534\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n",
      "(array([[0.9542289 , 0.07635854],\n",
      "       [0.96284026, 0.07434906],\n",
      "       [0.98706657, 0.03294165],\n",
      "       ...,\n",
      "       [0.912202  , 0.11904449],\n",
      "       [0.96240944, 0.07730153],\n",
      "       [0.9699621 , 0.06737549]], dtype=float32), 0.8468627233995463, (1666, 2))\n"
     ]
    }
   ],
   "source": [
    "#konvolucija s sekvenco in cobindingom\n",
    "\n",
    "for x in range (0,1):\n",
    "   \n",
    "    X = X \n",
    "    y = get_class(protein,\"train\")\n",
    "        \n",
    "    kf = KFold(n_splits=3) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 48,data_format='channels_last', input_shape=(101,27), strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(12, 2, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "        print(model.summary())  \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "    y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "    score = roc_auc_score(y_test, y_scores)      \n",
    "    print(y_scores, score, y_scores.shape)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "def load_data(path):\n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "def get_seq(protein, t_data):\n",
    "    \n",
    "    training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "    x_train = []\n",
    "    \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "        \n",
    "        for i in range(0,len(sequence)):\n",
    "            num_seq.append(nucleotide[sequence[i]])\n",
    "            \n",
    "        \n",
    "        x_train.append(num_seq)\n",
    "\n",
    "    x_train = np.array(x_train)\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "def get_class(protein, t_data):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"data/clip/%s/5000/test_sample_0\"% protein)\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "def get_cobinding(protein, t_data):\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "        cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "    \n",
    "    \n",
    "    cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.907360439618504, 0.8730726100364452, 0.9033539276257723, 0.8293810921287077, 0.8877277813299231, 0.868460362676301, 0.8737809871212685, 0.8856299163750961, 0.8491788391189787, 0.83964375]\n",
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "conv1d_122 (Conv1D)          (None, 54, 120)           155640    \n",
      "_________________________________________________________________\n",
      "max_pooling1d_61 (MaxPooling (None, 22, 120)           0         \n",
      "_________________________________________________________________\n",
      "conv1d_123 (Conv1D)          (None, 21, 12)            2892      \n",
      "_________________________________________________________________\n",
      "global_average_pooling1d_61  (None, 12)                0         \n",
      "_________________________________________________________________\n",
      "dense_121 (Dense)            (None, 200)               2600      \n",
      "_________________________________________________________________\n",
      "dense_122 (Dense)            (None, 2)                 402       \n",
      "=================================================================\n",
      "Total params: 161,534\n",
      "Trainable params: 161,534\n",
      "Non-trainable params: 0\n",
      "_________________________________________________________________\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "for x in range (0,1):\n",
    "    \n",
    "    protein = \"1_PARCLIP_AGO1234_hg19\"\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "        \n",
    "    kf = KFold(n_splits=10) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 48,data_format='channels_last', input_shape=(101,27), strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(12, 2, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))      \n",
    "        \n",
    "    print (score)\n",
    "    print(model.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8717589706030997"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "np.mean([0.907360439618504, 0.8730726100364452, 0.9033539276257723, 0.8293810921287077, 0.8877277813299231, 0.868460362676301, 0.8737809871212685, 0.8856299163750961, 0.8491788391189787, 0.83964375])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# one-hot encoding za sekvenco\n",
    "training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "x_train = np.zeros((5000,101,4))\n",
    "r = 0    \n",
    "    \n",
    "for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "    sequence = list(record.seq)                \n",
    "    nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "    num_seq = list() #sekvenca v številskem formatu\n",
    "     \n",
    "    \n",
    "    for i in range(0,len(sequence)):\n",
    "            num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "    X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "    for i in range (len(num_seq)):\n",
    "            if num_seq[i] <= 3:\n",
    "                X[:,i,num_seq[i]] = 1               \n",
    "               \n",
    "    x_train[r,:,:] = X\n",
    "    r = r + 1\n",
    "    \n",
    "\n",
    "\n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "def load_data(path):\n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "def get_seq(protein, t_data):\n",
    "    \n",
    "    training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "    x_train = np.zeros((5000,101,4))\n",
    "    r = 0    \n",
    "\n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v številskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "def get_class(protein, t_data):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"data/clip/%s/5000/test_sample_0\"% protein)\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "def get_cobinding(protein, t_data):\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "        cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "    \n",
    "    \n",
    "    cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_PARCLIP_AGO1234_hg19 finishhed\n",
      "2_PARCLIP_AGO2MNASE_hg19 finishhed\n",
      "3_HITSCLIP_Ago2_binding_clusters finishhed\n",
      "4_HITSCLIP_Ago2_binding_clusters_2 finishhed\n",
      "5_CLIPSEQ_AGO2_hg19 finishhed\n",
      "6_CLIP-seq-eIF4AIII_1 finishhed\n",
      "7_CLIP-seq-eIF4AIII_2 finishhed\n",
      "8_PARCLIP_ELAVL1_hg19 finishhed\n",
      "9_PARCLIP_ELAVL1MNASE_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "12_PARCLIP_EWSR1_hg19 finishhed\n",
      "13_PARCLIP_FUS_hg19 finishhed\n",
      "14_PARCLIP_FUS_mut_hg19 finishhed\n",
      "15_PARCLIP_IGF2BP123_hg19 finishhed\n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters finishhed\n",
      "17_ICLIP_HNRNPC_hg19 finishhed\n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome finishhed\n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome finishhed\n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome finishhed\n",
      "21_PARCLIP_MOV10_Sievers_hg19 finishhed\n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome finishhed\n",
      "23_PARCLIP_PUM2_hg19 finishhed\n",
      "24_PARCLIP_QKI_hg19 finishhed\n",
      "25_CLIPSEQ_SFRS1_hg19 finishhed\n",
      "26_PARCLIP_TAF15_hg19 finishhed\n",
      "27_ICLIP_TDP43_hg19 finishhed\n",
      "28_ICLIP_TIA1_hg19 finishhed\n",
      "29_ICLIP_TIAL1_hg19 finishhed\n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters finishhed\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters finishhed\n",
      "[0.8784274080116949, 0.712064476180941, 0.8264209383405812, 0.847009581737597, 0.6497208726529128, 0.9006135614110178, 0.9433431526968606, 0.9477383033044304, 0.677806730286637, 0.9183066453987854, 0.9097194725956174, 0.9290900434189812, 0.9147761168064094, 0.9806999568739023, 0.8520893736781467, 0.9251161851548252, 0.9609008323111736, 0.7690864626572951, 0.6814716176149236, 0.7182194403291076, 0.9351248802543546, 0.7921685512048391, 0.9658200903838713, 0.9428692308457892, 0.8707048992001145, 0.9701312315204162, 0.8487082760564476, 0.9368091889596142, 0.9132027079540356, 0.9172633094372301, 0.9127423695032235]\n"
     ]
    }
   ],
   "source": [
    "score_list = []\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "    \n",
    "    size = X.shape[2]\n",
    "    \n",
    "    kf = KFold(n_splits=10) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 24,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=24, strides=6, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "        \n",
    "    \n",
    "    \n",
    "    print (\"%s finishhed\" % (protein))\n",
    "    score_list.append(np.mean(score))\n",
    "    \n",
    "    \n",
    "print(score_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                                0.180427  0.878427   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.111064  0.712064   \n",
      "3_HITSCLIP_Ago2_binding_clusters                      0.065421  0.826421   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                    0.105010  0.847010   \n",
      "5_CLIPSEQ_AGO2_hg19                                   0.056721  0.649721   \n",
      "6_CLIP-seq-eIF4AIII_1                                 0.037614  0.900614   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.080343  0.943343   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.023738  0.947738   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                            0.083807  0.677807   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.041693  0.918307   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.046281  0.909719   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.066090  0.929090   \n",
      "13_PARCLIP_FUS_hg19                                   0.116776  0.914776   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.050700  0.980700   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.153089  0.852089   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.026884  0.925116   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.012099  0.960901   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.018914  0.769086   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.031528  0.681472   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...   -0.009781  0.718219   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.028125  0.935125   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.030831  0.792169   \n",
      "23_PARCLIP_PUM2_hg19                                  0.029820  0.965820   \n",
      "24_PARCLIP_QKI_hg19                                   0.041869  0.942869   \n",
      "25_CLIPSEQ_SFRS1_hg19                                 0.083705  0.870705   \n",
      "26_PARCLIP_TAF15_hg19                                 0.073131  0.970131   \n",
      "27_ICLIP_TDP43_hg19                                   0.115708  0.848708   \n",
      "28_ICLIP_TIA1_hg19                                    0.008809  0.936809   \n",
      "29_ICLIP_TIAL1_hg19                                   0.006203  0.913203   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters          0.099263  0.917263   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.117742  0.912742   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.698  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.601  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.761  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.742  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.593  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.863  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.863  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.924  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.594  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.960  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.956  \n",
      "12_PARCLIP_EWSR1_hg19                               0.863  \n",
      "13_PARCLIP_FUS_hg19                                 0.798  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.930  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.699  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.952  \n",
      "17_ICLIP_HNRNPC_hg19                                0.973  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.788  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.713  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.728  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.907  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.823  \n",
      "23_PARCLIP_PUM2_hg19                                0.936  \n",
      "24_PARCLIP_QKI_hg19                                 0.901  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.787  \n",
      "26_PARCLIP_TAF15_hg19                               0.897  \n",
      "27_ICLIP_TDP43_hg19                                 0.733  \n",
      "28_ICLIP_TIA1_hg19                                  0.928  \n",
      "29_ICLIP_TIAL1_hg19                                 0.907  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.818  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.795  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "score = [0.8784274080116949, 0.712064476180941, 0.8264209383405812, 0.847009581737597, 0.6497208726529128, 0.9006135614110178, 0.9433431526968606, 0.9477383033044304, 0.677806730286637, 0.9183066453987854, 0.9097194725956174, 0.9290900434189812, 0.9147761168064094, 0.9806999568739023, 0.8520893736781467, 0.9251161851548252, 0.9609008323111736, 0.7690864626572951, 0.6814716176149236, 0.7182194403291076, 0.9351248802543546, 0.7921685512048391, 0.9658200903838713, 0.9428692308457892, 0.8707048992001145, 0.9701312315204162, 0.8487082760564476, 0.9368091889596142, 0.9132027079540356, 0.9172633094372301, 0.9127423695032235]\n",
    "ionmf = [0.698, 0.601, 0.761, 0.742, 0.593, 0.863, 0.863, 0.924, 0.594, 0.960, 0.956, 0.863, 0.798, 0.930, 0.699, 0.952, 0.973, 0.788, 0.713, 0.728, 0.907, 0.823, 0.936, 0.901, 0.787, 0.897, 0.733, 0.928, 0.907, 0.818, 0.795]\n",
    "difference = np.subtract(score,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n",
    "\n",
    "\n",
    " \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_PARCLIP_AGO1234_hg19 finishhed\n",
      "2_PARCLIP_AGO2MNASE_hg19 finishhed\n",
      "3_HITSCLIP_Ago2_binding_clusters finishhed\n",
      "4_HITSCLIP_Ago2_binding_clusters_2 finishhed\n",
      "5_CLIPSEQ_AGO2_hg19 finishhed\n",
      "6_CLIP-seq-eIF4AIII_1 finishhed\n",
      "7_CLIP-seq-eIF4AIII_2 finishhed\n",
      "8_PARCLIP_ELAVL1_hg19 finishhed\n",
      "9_PARCLIP_ELAVL1MNASE_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "12_PARCLIP_EWSR1_hg19 finishhed\n",
      "13_PARCLIP_FUS_hg19 finishhed\n",
      "14_PARCLIP_FUS_mut_hg19 finishhed\n",
      "15_PARCLIP_IGF2BP123_hg19 finishhed\n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters finishhed\n",
      "17_ICLIP_HNRNPC_hg19 finishhed\n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome finishhed\n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome finishhed\n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome finishhed\n",
      "21_PARCLIP_MOV10_Sievers_hg19 finishhed\n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome finishhed\n",
      "23_PARCLIP_PUM2_hg19 finishhed\n",
      "24_PARCLIP_QKI_hg19 finishhed\n",
      "25_CLIPSEQ_SFRS1_hg19 finishhed\n",
      "26_PARCLIP_TAF15_hg19 finishhed\n",
      "27_ICLIP_TDP43_hg19 finishhed\n",
      "28_ICLIP_TIA1_hg19 finishhed\n",
      "29_ICLIP_TIAL1_hg19 finishhed\n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters finishhed\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters finishhed\n",
      "[0.8854467745426089, 0.7389716116566503, 0.8486800058647509, 0.8558456066418886, 0.7024250393464896, 0.8966307808447306, 0.935885864297013, 0.9493017292655587, 0.7186794497766076, 0.9292762804991401, 0.9173037219001344, 0.9206776938194293, 0.8934579091067907, 0.9738436234314399, 0.8785229371893042, 0.9172132931900181, 0.9489552102800551, 0.7735668951519872, 0.724478273126381, 0.7632085437513132, 0.9432564453494334, 0.8101318069278541, 0.9730036404398559, 0.9426504774343695, 0.8595774908446335, 0.9606016792315117, 0.858870168513284, 0.9204543898078441, 0.9140122739172712, 0.8805678751447241, 0.8895041219714479]\n"
     ]
    }
   ],
   "source": [
    "score_list = []\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "    \n",
    "    size = X.shape[2]\n",
    "    \n",
    "    kf = KFold(n_splits=10) #10 različnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 24,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "        \n",
    "    \n",
    "    \n",
    "    print (\"%s finishhed\" % (protein))\n",
    "    score_list.append(np.mean(score))\n",
    "    \n",
    "    \n",
    "print(score_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                                0.187447  0.885447   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.137972  0.738972   \n",
      "3_HITSCLIP_Ago2_binding_clusters                      0.087680  0.848680   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                    0.113846  0.855846   \n",
      "5_CLIPSEQ_AGO2_hg19                                   0.109425  0.702425   \n",
      "6_CLIP-seq-eIF4AIII_1                                 0.033631  0.896631   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.072886  0.935886   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.025302  0.949302   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                            0.124679  0.718679   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.030724  0.929276   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.038696  0.917304   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.057678  0.920678   \n",
      "13_PARCLIP_FUS_hg19                                   0.095458  0.893458   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.043844  0.973844   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.179523  0.878523   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.034787  0.917213   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.024045  0.948955   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.014433  0.773567   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...    0.011478  0.724478   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...    0.035209  0.763209   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.036256  0.943256   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.012868  0.810132   \n",
      "23_PARCLIP_PUM2_hg19                                  0.037004  0.973004   \n",
      "24_PARCLIP_QKI_hg19                                   0.041650  0.942650   \n",
      "25_CLIPSEQ_SFRS1_hg19                                 0.072577  0.859577   \n",
      "26_PARCLIP_TAF15_hg19                                 0.063602  0.960602   \n",
      "27_ICLIP_TDP43_hg19                                   0.125870  0.858870   \n",
      "28_ICLIP_TIA1_hg19                                   -0.007546  0.920454   \n",
      "29_ICLIP_TIAL1_hg19                                   0.007012  0.914012   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters          0.062568  0.880568   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.094504  0.889504   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.698  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.601  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.761  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.742  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.593  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.863  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.863  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.924  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.594  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.960  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.956  \n",
      "12_PARCLIP_EWSR1_hg19                               0.863  \n",
      "13_PARCLIP_FUS_hg19                                 0.798  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.930  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.699  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.952  \n",
      "17_ICLIP_HNRNPC_hg19                                0.973  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.788  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.713  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.728  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.907  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.823  \n",
      "23_PARCLIP_PUM2_hg19                                0.936  \n",
      "24_PARCLIP_QKI_hg19                                 0.901  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.787  \n",
      "26_PARCLIP_TAF15_hg19                               0.897  \n",
      "27_ICLIP_TDP43_hg19                                 0.733  \n",
      "28_ICLIP_TIA1_hg19                                  0.928  \n",
      "29_ICLIP_TIAL1_hg19                                 0.907  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.818  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.795  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "score = [0.8854467745426089, 0.7389716116566503, 0.8486800058647509, 0.8558456066418886, 0.7024250393464896, 0.8966307808447306, 0.935885864297013, 0.9493017292655587, 0.7186794497766076, 0.9292762804991401, 0.9173037219001344, 0.9206776938194293, 0.8934579091067907, 0.9738436234314399, 0.8785229371893042, 0.9172132931900181, 0.9489552102800551, 0.7735668951519872, 0.724478273126381, 0.7632085437513132, 0.9432564453494334, 0.8101318069278541, 0.9730036404398559, 0.9426504774343695, 0.8595774908446335, 0.9606016792315117, 0.858870168513284, 0.9204543898078441, 0.9140122739172712, 0.8805678751447241, 0.8895041219714479]\n",
    "ionmf = [0.698, 0.601, 0.761, 0.742, 0.593, 0.863, 0.863, 0.924, 0.594, 0.960, 0.956, 0.863, 0.798, 0.930, 0.699, 0.952, 0.973, 0.788, 0.713, 0.728, 0.907, 0.823, 0.936, 0.901, 0.787, 0.897, 0.733, 0.928, 0.907, 0.818, 0.795]\n",
    "difference = np.subtract(score,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
