{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "protein = \"1_PARCLIP_AGO1234_hg19\"\n",
    "\n",
    "with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_RNAfold.tab.gz\"% protein), \"rt\") as f:\n",
    "    fold_data = np.loadtxt(f, skiprows=1) \n",
    "\n",
    "\n",
    "    fold = np.zeros((5000,101,fold_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5000, 101, 1)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fold.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "\n",
    "def load_data(path):\n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data):\n",
    "    \n",
    "    training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "    x_train = np.zeros((5000,101,4))\n",
    "    r = 0    \n",
    "\n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v Å¡tevilskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"data/clip/%s/5000/test_sample_0\"% protein)\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data):\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "        cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "    \n",
    "    \n",
    "    cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data):\n",
    "\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_RegionType.tab.gz\"% protein), \"rt\") as f:\n",
    "        region_data = np.loadtxt(f, skiprows=1) \n",
    "\n",
    "\n",
    "    region = np.zeros((5000,101,region_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n",
    "\n",
    "\n",
    "def get_fold (protein, t_data):\n",
    "   \n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_RNAfold.tab.gz\"% protein), \"rt\") as f:\n",
    "        fold_data = np.loadtxt(f, skiprows=1) \n",
    "    \n",
    "    fold = np.zeros((5000,101,fold_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "    \n",
    "    \n",
    "    return fold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_PARCLIP_AGO1234_hg19 finishhed\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-8c7acd606aa6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     31\u001b[0m         \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtrain_index\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtest_index\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m         \u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_train\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbatch_size\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m16\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nfs/home/alovse/workspace38/ENV/local/lib/python2.7/site-packages/keras/engine/training.pyc\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, **kwargs)\u001b[0m\n\u001b[1;32m   1035\u001b[0m                                         \u001b[0minitial_epoch\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0minitial_epoch\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1036\u001b[0m                                         \u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1037\u001b[0;31m                                         validation_steps=validation_steps)\n\u001b[0m\u001b[1;32m   1038\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1039\u001b[0m     def evaluate(self, x=None, y=None,\n",
      "\u001b[0;32m/nfs/home/alovse/workspace38/ENV/local/lib/python2.7/site-packages/keras/engine/training_arrays.pyc\u001b[0m in \u001b[0;36mfit_loop\u001b[0;34m(model, f, ins, out_labels, batch_size, epochs, verbose, callbacks, val_f, val_ins, shuffle, callback_metrics, initial_epoch, steps_per_epoch, validation_steps)\u001b[0m\n\u001b[1;32m    197\u001b[0m                     \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 199\u001b[0;31m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    200\u001b[0m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nfs/home/alovse/workspace38/ENV/local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.pyc\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   2656\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2657\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__call__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2658\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_session\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'_make_callable_from_options'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2659\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mpy_any\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mis_sparse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2660\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mpy_any\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mis_tensor\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0minputs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/nfs/home/alovse/workspace38/ENV/local/lib/python2.7/site-packages/keras/backend/tensorflow_backend.pyc\u001b[0m in \u001b[0;36mget_session\u001b[0;34m()\u001b[0m\n\u001b[1;32m    202\u001b[0m                     \u001b[0mv\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_keras_initialized\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    203\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0muninitialized_vars\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 204\u001b[0;31m                     \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvariables_initializer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0muninitialized_vars\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    205\u001b[0m     \u001b[0;31m# hack for list_devices() function.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    206\u001b[0m     \u001b[0;31m# list_devices() function is not available under tensorflow r1.3.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python2.7/contextlib.pyc\u001b[0m in \u001b[0;36m__exit__\u001b[0;34m(self, type, value, traceback)\u001b[0m\n\u001b[1;32m     19\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"generator didn't yield\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m     \u001b[0;32mdef\u001b[0m \u001b[0m__exit__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraceback\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mtype\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "score_list = []\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\"), get_region(protein,\"train\"), get_fold(protein, \"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "    \n",
    "    size = X.shape[2]\n",
    "    \n",
    "    kf = KFold(n_splits=10) #10 razliÄnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 24,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=24, strides=6, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "        \n",
    "    \n",
    "    \n",
    "    print (\"%s finishhed\" % (protein))\n",
    "    score_list.append(np.mean(score))\n",
    "    \n",
    "    \n",
    "print(score_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                               -0.004069  0.922931   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.013386  0.737386   \n",
      "3_HITSCLIP_Ago2_binding_clusters                     -0.009497  0.918503   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                   -0.010348  0.921652   \n",
      "5_CLIPSEQ_AGO2_hg19                                  -0.030713  0.700287   \n",
      "6_CLIP-seq-eIF4AIII_1                                -0.025159  0.932841   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.004481  0.968481   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.013070  0.961070   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                           -0.063612  0.713388   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.042708  0.930292   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.039203  0.929797   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.036053  0.930053   \n",
      "13_PARCLIP_FUS_hg19                                   0.079614  0.908614   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.026609  0.980609   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.001153  0.945153   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.018456  0.933544   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.012195  0.962805   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.038225  0.783775   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.124399  0.665601   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...   -0.039133  0.761867   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.009468  0.968468   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.026492  0.842508   \n",
      "23_PARCLIP_PUM2_hg19                                  0.021894  0.977894   \n",
      "24_PARCLIP_QKI_hg19                                   0.081032  0.949032   \n",
      "25_CLIPSEQ_SFRS1_hg19                                -0.033459  0.879541   \n",
      "26_PARCLIP_TAF15_hg19                                 0.074034  0.974034   \n",
      "27_ICLIP_TDP43_hg19                                   0.011773  0.869773   \n",
      "28_ICLIP_TIA1_hg19                                   -0.015249  0.933751   \n",
      "29_ICLIP_TIAL1_hg19                                   0.001529  0.932529   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters         -0.001926  0.931074   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.032216  0.936216   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.927  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.724  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.928  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.932  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.731  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.958  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.964  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.948  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.777  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.973  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.969  \n",
      "12_PARCLIP_EWSR1_hg19                               0.894  \n",
      "13_PARCLIP_FUS_hg19                                 0.829  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.954  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.944  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.952  \n",
      "17_ICLIP_HNRNPC_hg19                                0.975  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.822  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.790  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.801  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.959  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.869  \n",
      "23_PARCLIP_PUM2_hg19                                0.956  \n",
      "24_PARCLIP_QKI_hg19                                 0.868  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.913  \n",
      "26_PARCLIP_TAF15_hg19                               0.900  \n",
      "27_ICLIP_TDP43_hg19                                 0.858  \n",
      "28_ICLIP_TIA1_hg19                                  0.949  \n",
      "29_ICLIP_TIAL1_hg19                                 0.931  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.933  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.904  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "ionmf = [0.927, 0.724, 0.928, 0.932, 0.731, 0.958, 0.964, 0.948, 0.777, 0.973, 0.969, 0.894, 0.829, 0.954, 0.944, 0.952, 0.975, 0.822, 0.790, 0.801, 0.959, 0.869, 0.956, 0.868, 0.913, 0.900, 0.858, 0.949, 0.931, 0.933, 0.904]\n",
    "score = [0.9229308461970577, 0.7373859595303559, 0.9185030153657214, 0.9216517570803745, 0.7002868366836459, 0.9328408367848035, 0.968480765443396, 0.9610704233808134, 0.7133883127703811, 0.930291845973384, 0.9297965698375279, 0.9300525415845116, 0.908614293424964, 0.980608867261066, 0.9451526335195901, 0.933543787357689, 0.9628050181895601, 0.7837749438422019, 0.6656009182246273, 0.7618671987601008, 0.9684676359870534, 0.8425083385398459, 0.9778942059708827, 0.9490315944769474, 0.8795407520897189, 0.9740336879062095, 0.8697733106206457, 0.9337510003903715, 0.9325289769216644, 0.9310742382453737, 0.9362160341422259]\n",
    "\n",
    "difference = np.subtract(score,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "\n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\"), get_region(protein,\"train\"), get_fold(protein, \"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "\n",
    "    size = X.shape[2]\n",
    "\n",
    "    kf = KFold(n_splits=10) #10 razliÄnih delitev podatkov\n",
    "\n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "\n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120,4 ,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=24, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "\n",
    "\n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "\n",
    "\n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "\n",
    "\n",
    "\n",
    "    print (\"%s finishhed\" % (protein))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                                0.000228  0.927228   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.045494  0.769494   \n",
      "3_HITSCLIP_Ago2_binding_clusters                      0.003355  0.931355   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                   -0.002395  0.929605   \n",
      "5_CLIPSEQ_AGO2_hg19                                  -0.001726  0.729274   \n",
      "6_CLIP-seq-eIF4AIII_1                                -0.021004  0.936996   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.002070  0.966070   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.015484  0.963484   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                           -0.021286  0.755714   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.039387  0.933613   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.038586  0.930414   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.036239  0.930239   \n",
      "13_PARCLIP_FUS_hg19                                   0.067760  0.896760   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.023635  0.977635   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.002659  0.946659   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.026051  0.925949   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.025809  0.949191   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.015235  0.806765   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.022257  0.767743   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...    0.003702  0.804702   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.014330  0.973330   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.007606  0.861394   \n",
      "23_PARCLIP_PUM2_hg19                                  0.020779  0.976779   \n",
      "24_PARCLIP_QKI_hg19                                   0.078283  0.946283   \n",
      "25_CLIPSEQ_SFRS1_hg19                                -0.020999  0.892001   \n",
      "26_PARCLIP_TAF15_hg19                                 0.060460  0.960460   \n",
      "27_ICLIP_TDP43_hg19                                   0.017196  0.875196   \n",
      "28_ICLIP_TIA1_hg19                                   -0.017483  0.931517   \n",
      "29_ICLIP_TIAL1_hg19                                  -0.001047  0.929953   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters         -0.025991  0.907009   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.023670  0.927670   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.927  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.724  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.928  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.932  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.731  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.958  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.964  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.948  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.777  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.973  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.969  \n",
      "12_PARCLIP_EWSR1_hg19                               0.894  \n",
      "13_PARCLIP_FUS_hg19                                 0.829  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.954  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.944  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.952  \n",
      "17_ICLIP_HNRNPC_hg19                                0.975  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.822  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.790  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.801  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.959  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.869  \n",
      "23_PARCLIP_PUM2_hg19                                0.956  \n",
      "24_PARCLIP_QKI_hg19                                 0.868  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.913  \n",
      "26_PARCLIP_TAF15_hg19                               0.900  \n",
      "27_ICLIP_TDP43_hg19                                 0.858  \n",
      "28_ICLIP_TIA1_hg19                                  0.949  \n",
      "29_ICLIP_TIAL1_hg19                                 0.931  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.933  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.904  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "ionmf = [0.927, 0.724, 0.928, 0.932, 0.731, 0.958, 0.964, 0.948, 0.777, 0.973, 0.969, 0.894, 0.829, 0.954, 0.944, 0.952, 0.975, 0.822, 0.790, 0.801, 0.959, 0.869, 0.956, 0.868, 0.913, 0.900, 0.858, 0.949, 0.931, 0.933, 0.904]\n",
    "\n",
    "score = [0.927228146855331, 0.769494283541073, 0.9313552127427794, 0.9296050749854965, 0.7292741564727823, 0.9369963374903605, 0.9660701142482843, 0.9634841972960959, 0.7557140532300421, 0.933612511482227, 0.9304135125615302, 0.9302392163625471, 0.8967602361845829, 0.9776346381307371, 0.946658674319971, 0.9259490941528631, 0.9491907567623634, 0.8067653522243997, 0.7677425192750433, 0.8047021405346483, 0.9733297070917741, 0.8613936257163806, 0.9767794248136615, 0.9462828423885125, 0.8920009103049378, 0.960460082563338, 0.875195852050852, 0.9315173833880024, 0.9299525323391871, 0.9070087940631355, 0.9276696990129027]\n",
    "\n",
    "\n",
    "difference = np.subtract(score,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
