{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Training the model on 30k datasets.\n",
    "\n",
    "import random\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "import pickle\n",
    "from tensorflow import set_random_seed\n",
    "from keras.models import Sequential\n",
    "from keras.callbacks import ModelCheckpoint, EarlyStopping\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Set random seeds for reproducibility.\n",
    "np.random.seed(4)\n",
    "random.seed(5)\n",
    "set_random_seed(4) \n",
    "\n",
    "\n",
    "\n",
    "def load_data(path):   \n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data, training_set_number): \n",
    "    if t_data == \"train\":\n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((30000,101,4))          \n",
    "    \n",
    "    elif t_data == \"test\":    \n",
    "                \n",
    "        training_data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "        x_train = np.zeros((10000,101,4))      \n",
    "      \n",
    "    r = 0    \n",
    "      \n",
    "    \n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v Å¡tevilskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        \n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data,training_set_number):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"datasets/clip/%s/30000/training_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"datasets/clip/%s/30000/test_sample_%s\"% (protein, training_set_number))\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        # [1,0] if there was no observed binding and [0,1] for sequences where binding was observed.\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "            \n",
    "        cobinding = np.zeros((30000,101,cobinding_data.shape[1]/101),dtype=np.int)    \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_Cobinding.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "        cobinding = np.zeros((10000,101,cobinding_data.shape[1]/101),dtype=np.int)\n",
    "   \n",
    "    \n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1)\n",
    "        region = np.zeros((30000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RegionType.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            region_data = np.loadtxt(f, skiprows=1) \n",
    "        region = np.zeros((10000,101,region_data.shape[1]/101),dtype=np.int)\n",
    "    \n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n",
    "\n",
    "\n",
    "def get_fold (protein, t_data, training_set_number):\n",
    "    if t_data == \"train\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/training_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((30000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "                             \n",
    "    elif t_data == \"test\":\n",
    "        with gzip.open((\"datasets/clip/%s/30000/test_sample_%s/matrix_RNAfold.tab.gz\"% (protein, training_set_number)), \"rt\") as f:\n",
    "            fold_data = np.loadtxt(f, skiprows=1) \n",
    "        fold = np.zeros((10000,101,fold_data.shape[1]/101),dtype=np.int)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    for n in range(0,fold_data.shape[1],101):\n",
    "        a = fold_data[:,n:(n+101)]\n",
    "        fold[:,:,(n/101)] = a\n",
    "    \n",
    "    \n",
    "    return fold\n",
    "\n",
    "def load_data_sources(protein, t_data, training_set_number, *args):\n",
    "    X = np.array([])\n",
    "    data_sources = []\n",
    "    for arg in args:\n",
    "        \n",
    "        if arg == 'KMER':\n",
    "            if X.size == 0:\n",
    "                X = get_seq(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_seq(protein, t_data, training_set_number)))\n",
    "        if arg == 'RNA': \n",
    "            if X.size == 0:\n",
    "                X = get_fold(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_fold(protein, t_data, training_set_number)))\n",
    "        if arg == 'RG':   \n",
    "            if X.size == 0:\n",
    "                X = get_region(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_region(protein, t_data, training_set_number)))\n",
    "        if arg == 'CLIP': \n",
    "            if X.size == 0:\n",
    "                X = get_cobinding(protein, t_data, training_set_number)\n",
    "            else: \n",
    "                X = np.dstack((X, get_cobinding(protein, t_data, training_set_number)))\n",
    "        data_sources.append(arg)\n",
    "        \n",
    "    data_sources = ','.join(data_sources)\n",
    "    return data_sources, X\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from vis.visualization import visualize_saliency\n",
    "from vis.utils import utils\n",
    "from keras import activations\n",
    "from keras.models import model_from_json\n",
    "\n",
    "# Sum of gradients across all layers saved under protein name.\n",
    "gradients_dict = {}\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "experiment_set = ['KMER' , 'RNA', 'RG', 'CLIP']\n",
    "\n",
    "for training_set_number in range (1):\n",
    "    for protein in protein_list:\n",
    "        print (protein)\n",
    "        \n",
    "        \n",
    "        #Load the model and weights.\n",
    "        with open(\"results/set_%s/%s/model.json\" % (training_set_number, protein), \"r\") as json_file:\n",
    "            json = json_file.read()\n",
    "            model = model_from_json(json)\n",
    "            model.load_weights(\"results/set_%s/%s/weights.h5\" % (training_set_number, protein))\n",
    "\n",
    "\n",
    "        #Load data for testing purposes.\n",
    "        data_sources, X_test = load_data_sources(protein, 'test', training_set_number, *experiment_set)\n",
    "        \n",
    "\n",
    "        layer_index = 0\n",
    "        protein_grads = np.zeros((12,101))\n",
    "\n",
    "        for layer in model.layers:\n",
    "            print (layer)\n",
    "\n",
    "            #Swap softmax with linear\n",
    "            model.layers[layer_index].activation = activations.linear\n",
    "            model = utils.apply_modifications(model)\n",
    "\n",
    "            grads = visualize_saliency(model, layer_index, filter_indices=1, seed_input=X_test)\n",
    "\n",
    "            protein_grads[layer_index,:] = grads\n",
    "\n",
    "            plt.figure(1)\n",
    "            plt.subplot(12,1,layer_index+1, xticks = np.arange(-50,50,10), yticks = [])    \n",
    "            plt.imshow(grads.reshape((1, -1)), aspect='auto', cmap='jet', interpolation='nearest')\n",
    "\n",
    "            layer_index += 1\n",
    "\n",
    "        plt.figure(1)\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "        \n",
    "        grads_total = np.sum(protein_grads, axis = 0)\n",
    "        gradients_dict[protein] = grads_total\n",
    "        \n",
    "        plt.figure(2)\n",
    "        plt.subplot(211, xticks=np.arange(0,101,10), yticks=[])\n",
    "        plt.imshow(grads_total.reshape((1,-1)), aspect='auto', cmap='jet', interpolation='nearest')\n",
    "        plt.show()\n",
    "        plt.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "gradient_array = []\n",
    "for protein in protein_list:\n",
    "    gradient_array.append(gradients_dict.get(protein))\n",
    "\n",
    "Z = linkage(gradient_array, 'ward')\n",
    "fig = plt.figure(figsize=(10, 10))\n",
    "dn = dendrogram(Z, orientation = 'left')\n",
    "plt.show()\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
