{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "\n",
    "np.random.seed(454)\n",
    "\n",
    "\n",
    "def load_data(path):\n",
    "    data = gzip.open(os.path.join(path,\"sequences.fa.gz\"),\"rt\")\n",
    "    return data\n",
    "\n",
    "\n",
    "def get_seq(protein, t_data):\n",
    "    \n",
    "    training_data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "    x_train = np.zeros((5000,101,4))\n",
    "    r = 0    \n",
    "\n",
    "    for record in SeqIO.parse(training_data,\"fasta\"):\n",
    "        sequence = list(record.seq)                \n",
    "        nucleotide = {'A' : 0, 'T' : 1, 'G' : 2, 'C' : 3, 'N' : 4} \n",
    "        num_seq = list() #sekvenca v Å¡tevilskem formatu\n",
    "\n",
    "\n",
    "        for i in range(0,len(sequence)):\n",
    "                num_seq.append(nucleotide[sequence[i]])\n",
    "\n",
    "\n",
    "        X = np.zeros((1,len(num_seq),4))\n",
    "\n",
    "        for i in range (len(num_seq)):\n",
    "                if num_seq[i] <= 3:\n",
    "                    X[:,i,num_seq[i]] = 1               \n",
    "\n",
    "        x_train[r,:,:] = X\n",
    "        r = r + 1\n",
    "    \n",
    "    return x_train\n",
    "\n",
    "\n",
    "def get_class(protein, t_data):\n",
    "    y_train = []\n",
    "    \n",
    "\n",
    "    if t_data == 'train':\n",
    "        data = load_data(\"data/clip/%s/5000/training_sample_0\"% protein)\n",
    "\n",
    "    elif t_data == 'test':\n",
    "        data = load_data(\"data/clip/%s/5000/test_sample_0\"% protein)\n",
    "\n",
    "\n",
    "    for record in SeqIO.parse(data,\"fasta\"):\n",
    "        v = int((record.description).split(\":\")[1])\n",
    "        y_train.append([int(v == 0), int(v != 0)])\n",
    "\n",
    "    y_train = np.array(y_train)\n",
    "    return y_train\n",
    "\n",
    "\n",
    "def get_cobinding(protein, t_data):\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_Cobinding.tab.gz\"% protein), \"rt\") as f:\n",
    "        cobinding_data = np.loadtxt(f, skiprows=1) \n",
    "    \n",
    "    \n",
    "    cobinding = np.zeros((5000,101,cobinding_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,cobinding_data.shape[1],101):\n",
    "        a = cobinding_data[:,n:(n+101)]\n",
    "        cobinding[:,:,(n/101)] = a\n",
    "    \n",
    "    return cobinding\n",
    "    \n",
    "\n",
    "def get_region (protein, t_data):\n",
    "\n",
    "    with gzip.open((\"data/clip/%s/5000/training_sample_0/matrix_RegionType.tab.gz\"% protein), \"rt\") as f:\n",
    "        region_data = np.loadtxt(f, skiprows=1) \n",
    "\n",
    "\n",
    "    region = np.zeros((5000,101,region_data.shape[1]/101),dtype=np.int) #ustvarim prazen array\n",
    "\n",
    "\n",
    "    for n in range(0,region_data.shape[1],101):\n",
    "        a = region_data[:,n:(n+101)]\n",
    "        region[:,:,(n/101)] = a\n",
    "        \n",
    "    return region\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1_PARCLIP_AGO1234_hg19 finishhed\n",
      "2_PARCLIP_AGO2MNASE_hg19 finishhed\n",
      "3_HITSCLIP_Ago2_binding_clusters finishhed\n",
      "4_HITSCLIP_Ago2_binding_clusters_2 finishhed\n",
      "5_CLIPSEQ_AGO2_hg19 finishhed\n",
      "6_CLIP-seq-eIF4AIII_1 finishhed\n",
      "7_CLIP-seq-eIF4AIII_2 finishhed\n",
      "8_PARCLIP_ELAVL1_hg19 finishhed\n",
      "9_PARCLIP_ELAVL1MNASE_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "10_PARCLIP_ELAVL1A_hg19 finishhed\n",
      "12_PARCLIP_EWSR1_hg19 finishhed\n",
      "13_PARCLIP_FUS_hg19 finishhed\n",
      "14_PARCLIP_FUS_mut_hg19 finishhed\n",
      "15_PARCLIP_IGF2BP123_hg19 finishhed\n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters finishhed\n",
      "17_ICLIP_HNRNPC_hg19 finishhed\n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome finishhed\n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome finishhed\n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome finishhed\n",
      "21_PARCLIP_MOV10_Sievers_hg19 finishhed\n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome finishhed\n",
      "23_PARCLIP_PUM2_hg19 finishhed\n",
      "24_PARCLIP_QKI_hg19 finishhed\n",
      "25_CLIPSEQ_SFRS1_hg19 finishhed\n",
      "26_PARCLIP_TAF15_hg19 finishhed\n",
      "27_ICLIP_TDP43_hg19 finishhed\n",
      "28_ICLIP_TIA1_hg19 finishhed\n",
      "29_ICLIP_TIAL1_hg19 finishhed\n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters finishhed\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters finishhed\n",
      "[0.9122606727849598, 0.7352028341204624, 0.9148904726624577, 0.9140625368320323, 0.7066025908174656, 0.9197919208547927, 0.9524650562914404, 0.9622171914911885, 0.7101035456143489, 0.9285236030684215, 0.9306464453582949, 0.933199140729848, 0.9112582796030366, 0.9816859564997011, 0.9413768897897411, 0.9309535920027695, 0.9642977128360997, 0.7952061246423778, 0.7037769535155458, 0.7592575829101404, 0.9696535997173136, 0.8310232375581794, 0.9760468727007974, 0.9533679999498347, 0.8750664026525394, 0.9762533229844822, 0.8776657040368716, 0.9378902495507401, 0.9295940146136404, 0.9351547816201979, 0.9364341930141972]\n",
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                               -0.000739  0.912261   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.021203  0.735203   \n",
      "3_HITSCLIP_Ago2_binding_clusters                     -0.013110  0.914890   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                   -0.010937  0.914063   \n",
      "5_CLIPSEQ_AGO2_hg19                                  -0.045397  0.706603   \n",
      "6_CLIP-seq-eIF4AIII_1                                -0.019208  0.919792   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.010465  0.952465   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.008217  0.962217   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                           -0.012896  0.710104   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.045476  0.928524   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.040354  0.930646   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.040199  0.933199   \n",
      "13_PARCLIP_FUS_hg19                                   0.100258  0.911258   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.026686  0.981686   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.005377  0.941377   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.011046  0.930954   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.011702  0.964298   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.030794  0.795206   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.097223  0.703777   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...   -0.036742  0.759258   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.013654  0.969654   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.036977  0.831023   \n",
      "23_PARCLIP_PUM2_hg19                                  0.015047  0.976047   \n",
      "24_PARCLIP_QKI_hg19                                   0.076368  0.953368   \n",
      "25_CLIPSEQ_SFRS1_hg19                                -0.022934  0.875066   \n",
      "26_PARCLIP_TAF15_hg19                                 0.066253  0.976253   \n",
      "27_ICLIP_TDP43_hg19                                   0.017666  0.877666   \n",
      "28_ICLIP_TIA1_hg19                                   -0.005110  0.937890   \n",
      "29_ICLIP_TIAL1_hg19                                  -0.012406  0.929594   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters          0.020155  0.935155   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.011434  0.936434   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.913  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.714  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.928  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.925  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.752  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.939  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.942  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.954  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.723  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.974  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.971  \n",
      "12_PARCLIP_EWSR1_hg19                               0.893  \n",
      "13_PARCLIP_FUS_hg19                                 0.811  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.955  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.936  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.942  \n",
      "17_ICLIP_HNRNPC_hg19                                0.976  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.826  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.801  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.796  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.956  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.868  \n",
      "23_PARCLIP_PUM2_hg19                                0.961  \n",
      "24_PARCLIP_QKI_hg19                                 0.877  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.898  \n",
      "26_PARCLIP_TAF15_hg19                               0.910  \n",
      "27_ICLIP_TDP43_hg19                                 0.860  \n",
      "28_ICLIP_TIA1_hg19                                  0.943  \n",
      "29_ICLIP_TIAL1_hg19                                 0.942  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.915  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.925  \n"
     ]
    }
   ],
   "source": [
    "score_list = []\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\"), get_region(protein,\"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "    \n",
    "    size = X.shape[2]\n",
    "    \n",
    "    kf = KFold(n_splits=10) #10 razliÄnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 24,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=24, strides=6, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "        \n",
    "    \n",
    "    \n",
    "    print (\"%s finishhed\" % (protein))\n",
    "    score_list.append(np.mean(score))\n",
    "    \n",
    "    \n",
    "print(score_list)\n",
    "\n",
    "\n",
    "\n",
    "# Table for comparing results\n",
    "ionmf = [0.913, 0.714, 0.928, 0.925, 0.752, 0.939, 0.942, 0.954, 0.723, 0.974, 0.971, 0.893, 0.811, 0.955, 0.936, 0.942, 0.976, 0.826, 0.801, 0.796, 0.956, 0.868, 0.961, 0.877, 0.898, 0.910, 0.860, 0.943, 0.942, 0.915, 0.925]\n",
    "difference = np.subtract(score_list,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score_list, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "score_list = []\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "\n",
    "for x in range (0,31):\n",
    "\n",
    "    protein = protein_list[x]\n",
    "    \n",
    "    X = np.dstack((get_cobinding(protein,\"train\"), get_seq(protein,\"train\"), get_region(protein,\"train\")))\n",
    "    y = get_class(protein,\"train\")\n",
    "    \n",
    "    size = X.shape[2]\n",
    "    \n",
    "    kf = KFold(n_splits=10) #10 razliÄnih delitev podatkov  \n",
    "    \n",
    "    score = []\n",
    "    for train_index, test_index in kf.split(X):\n",
    "        \n",
    "        # sestavljanje modela\n",
    "        model = Sequential()\n",
    "        model.add(Conv1D(120, 24,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=12, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "          \n",
    "        \n",
    "        X_train, X_test = X[train_index], X[test_index]\n",
    "        y_train, y_test = y[train_index], y[test_index]\n",
    "\n",
    "        model.fit(X_train, y_train, epochs=10, batch_size=16, verbose=0)\n",
    "   \n",
    "    \n",
    "\n",
    "        y_scores =(model.predict(X_test))# pridobim napovedi za sekvence iz cobinding\n",
    "        score.append(roc_auc_score(y_test, y_scores))\n",
    "        \n",
    "    \n",
    "    \n",
    "    print (\"%s finishhed\" % (protein))\n",
    "    score_list.append(np.mean(score))\n",
    "    \n",
    "    \n",
    "print(score_list)\n",
    "\n",
    "\n",
    "\n",
    "# Table for comparing results\n",
    "ionmf = [0.913, 0.714, 0.928, 0.925, 0.752, 0.939, 0.942, 0.954, 0.723, 0.974, 0.971, 0.893, 0.811, 0.955, 0.936, 0.942, 0.976, 0.826, 0.801, 0.796, 0.956, 0.868, 0.961, 0.877, 0.898, 0.910, 0.860, 0.943, 0.942, 0.915, 0.925]\n",
    "difference = np.subtract(score_list,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score_list, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score  \\\n",
      "1_PARCLIP_AGO1234_hg19                                0.003631  0.916631   \n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.021584  0.735584   \n",
      "3_HITSCLIP_Ago2_binding_clusters                     -0.012313  0.915687   \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                   -0.014000  0.911000   \n",
      "5_CLIPSEQ_AGO2_hg19                                  -0.041303  0.710697   \n",
      "6_CLIP-seq-eIF4AIII_1                                -0.021415  0.917585   \n",
      "7_CLIP-seq-eIF4AIII_2                                 0.011200  0.953200   \n",
      "8_PARCLIP_ELAVL1_hg19                                 0.006934  0.960934   \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                           -0.005456  0.717544   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.047998  0.926002   \n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.040008  0.930992   \n",
      "12_PARCLIP_EWSR1_hg19                                 0.039382  0.932382   \n",
      "13_PARCLIP_FUS_hg19                                   0.096185  0.907185   \n",
      "14_PARCLIP_FUS_mut_hg19                               0.028140  0.983140   \n",
      "15_PARCLIP_IGF2BP123_hg19                             0.007377  0.943377   \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.007333  0.934667   \n",
      "17_ICLIP_HNRNPC_hg19                                 -0.016689  0.959311   \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.033496  0.792504   \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.094130  0.706870   \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...   -0.027814  0.768186   \n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.013496  0.969496   \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.055675  0.812325   \n",
      "23_PARCLIP_PUM2_hg19                                  0.016941  0.977941   \n",
      "24_PARCLIP_QKI_hg19                                   0.069942  0.946942   \n",
      "25_CLIPSEQ_SFRS1_hg19                                -0.027627  0.870373   \n",
      "26_PARCLIP_TAF15_hg19                                 0.060211  0.970211   \n",
      "27_ICLIP_TDP43_hg19                                   0.018139  0.878139   \n",
      "28_ICLIP_TIA1_hg19                                   -0.004296  0.938704   \n",
      "29_ICLIP_TIAL1_hg19                                  -0.004352  0.937648   \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters          0.018512  0.933512   \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters       0.009115  0.934115   \n",
      "\n",
      "                                                    iONMF  \n",
      "1_PARCLIP_AGO1234_hg19                              0.913  \n",
      "2_PARCLIP_AGO2MNASE_hg19                            0.714  \n",
      "3_HITSCLIP_Ago2_binding_clusters                    0.928  \n",
      "4_HITSCLIP_Ago2_binding_clusters_2                  0.925  \n",
      "5_CLIPSEQ_AGO2_hg19                                 0.752  \n",
      "6_CLIP-seq-eIF4AIII_1                               0.939  \n",
      "7_CLIP-seq-eIF4AIII_2                               0.942  \n",
      "8_PARCLIP_ELAVL1_hg19                               0.954  \n",
      "9_PARCLIP_ELAVL1MNASE_hg19                          0.723  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.974  \n",
      "10_PARCLIP_ELAVL1A_hg19                             0.971  \n",
      "12_PARCLIP_EWSR1_hg19                               0.893  \n",
      "13_PARCLIP_FUS_hg19                                 0.811  \n",
      "14_PARCLIP_FUS_mut_hg19                             0.955  \n",
      "15_PARCLIP_IGF2BP123_hg19                           0.936  \n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters             0.942  \n",
      "17_ICLIP_HNRNPC_hg19                                0.976  \n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...  0.826  \n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...  0.801  \n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...  0.796  \n",
      "21_PARCLIP_MOV10_Sievers_hg19                       0.956  \n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...  0.868  \n",
      "23_PARCLIP_PUM2_hg19                                0.961  \n",
      "24_PARCLIP_QKI_hg19                                 0.877  \n",
      "25_CLIPSEQ_SFRS1_hg19                               0.898  \n",
      "26_PARCLIP_TAF15_hg19                               0.910  \n",
      "27_ICLIP_TDP43_hg19                                 0.860  \n",
      "28_ICLIP_TIA1_hg19                                  0.943  \n",
      "29_ICLIP_TIAL1_hg19                                 0.942  \n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters        0.915  \n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters     0.925  \n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import os\n",
    "from keras.models import Sequential\n",
    "from keras.layers import Dense, Activation, Conv1D, MaxPooling1D, Flatten, GlobalAveragePooling1D, Dropout\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.metrics import roc_auc_score, roc_curve\n",
    "import pandas as pd\n",
    "\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "score_list= [0.9166314660608545, 0.7355842739040614, 0.9156874064729882, 0.9110003605733115, 0.7106967639824682, 0.9175852831044603, 0.9532003478242389, 0.9609342519839978, 0.7175438955225013, 0.9260018057724553, 0.9309924438043105, 0.9323816311487496, 0.9071851247804773, 0.9831397242073813, 0.9433773325676775, 0.9346670168149185, 0.9593113189919608, 0.7925042050312222, 0.7068700113148783, 0.7681858296389412, 0.969496120005114, 0.812325268153552, 0.9779414360292946, 0.9469422043878664, 0.8703725801929352, 0.9702106147520017, 0.8781391345632541, 0.9387041550637198, 0.9376481211314986, 0.9335122789612825, 0.9341154218778888]\n",
    "ionmf = [0.913, 0.714, 0.928, 0.925, 0.752, 0.939, 0.942, 0.954, 0.723, 0.974, 0.971, 0.893, 0.811, 0.955, 0.936, 0.942, 0.976, 0.826, 0.801, 0.796, 0.956, 0.868, 0.961, 0.877, 0.898, 0.910, 0.860, 0.943, 0.942, 0.915, 0.925]\n",
    "difference = np.subtract(score_list,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score_list, 'iONMF':ionmf, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    Difference     Score\n",
      "1_PARCLIP_AGO1234_hg19                               -0.000502  0.912498\n",
      "2_PARCLIP_AGO2MNASE_hg19                              0.048645  0.762645\n",
      "3_HITSCLIP_Ago2_binding_clusters                     -0.019549  0.908451\n",
      "4_HITSCLIP_Ago2_binding_clusters_2                   -0.005313  0.919687\n",
      "5_CLIPSEQ_AGO2_hg19                                  -0.018584  0.733416\n",
      "6_CLIP-seq-eIF4AIII_1                                -0.021666  0.917334\n",
      "7_CLIP-seq-eIF4AIII_2                                 0.013430  0.955430\n",
      "8_PARCLIP_ELAVL1_hg19                                 0.010370  0.964370\n",
      "9_PARCLIP_ELAVL1MNASE_hg19                            0.031857  0.754857\n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.043822  0.930178\n",
      "10_PARCLIP_ELAVL1A_hg19                              -0.040111  0.930889\n",
      "12_PARCLIP_EWSR1_hg19                                 0.039926  0.932926\n",
      "13_PARCLIP_FUS_hg19                                   0.087857  0.898857\n",
      "14_PARCLIP_FUS_mut_hg19                               0.021241  0.976241\n",
      "15_PARCLIP_IGF2BP123_hg19                             0.007355  0.943355\n",
      "16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters              -0.014844  0.927156\n",
      "17_ICLIP_HNRNPC_hg19                                 -0.033829  0.942171\n",
      "18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela...   -0.014834  0.811166\n",
      "19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266...   -0.035122  0.765878\n",
      "20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLli...    0.003814  0.799814\n",
      "21_PARCLIP_MOV10_Sievers_hg19                         0.018531  0.974531\n",
      "22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg1...   -0.017063  0.850937\n",
      "23_PARCLIP_PUM2_hg19                                  0.015889  0.976889\n",
      "24_PARCLIP_QKI_hg19                                   0.068753  0.945753\n",
      "25_CLIPSEQ_SFRS1_hg19                                -0.024746  0.873254\n",
      "26_PARCLIP_TAF15_hg19                                 0.047479  0.957479\n",
      "27_ICLIP_TDP43_hg19                                   0.019387  0.879387\n",
      "28_ICLIP_TIA1_hg19                                   -0.005530  0.937470\n",
      "29_ICLIP_TIAL1_hg19                                  -0.014391  0.927609\n",
      "30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters          0.000010  0.915010\n",
      "31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters      -0.001593  0.923407\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "arhitektura modela \n",
    " \n",
    "model = Sequential()\n",
    "        model.add(Conv1D(120,4,data_format='channels_last', input_shape=(101, size) , strides = 1, padding='valid'))\n",
    "        model.add(MaxPooling1D(pool_size=24, strides=2, padding='valid'))\n",
    "        model.add(Conv1D(120, 4, activation='relu'))\n",
    "        model.add(GlobalAveragePooling1D())\n",
    "        model.add(Dense(200, activation='relu'))\n",
    "        model.add(Dense(2, activation='sigmoid'))\n",
    "        model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "\n",
    "'''\n",
    "\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "score_list = [0.91249816495929, 0.7626450772807127, 0.9084510844575933, 0.9196874015687442, 0.7334162319501186, 0.9173341246812802, 0.9554295878289565, 0.9643697034835925, 0.7548565940975945, 0.9301782951531139, 0.9308885650545436, 0.932926330425737, 0.898857053374814, 0.976241435704717, 0.9433553857366352, 0.9271558289215133, 0.94217081786666, 0.81116644993974, 0.765878379881905, 0.7998138612249177, 0.9745312431338725, 0.8509365580449263, 0.976889009288503, 0.945752770259, 0.8732536961307217, 0.9574785004610682, 0.8793866943231377, 0.9374699429100511, 0.9276089090165114, 0.91500985610617, 0.9234073665024163]\n",
    "protein_list = [\"1_PARCLIP_AGO1234_hg19\", \"2_PARCLIP_AGO2MNASE_hg19\",\"3_HITSCLIP_Ago2_binding_clusters\",\"4_HITSCLIP_Ago2_binding_clusters_2\",\"5_CLIPSEQ_AGO2_hg19\", \"6_CLIP-seq-eIF4AIII_1\",\"7_CLIP-seq-eIF4AIII_2\",\"8_PARCLIP_ELAVL1_hg19\",\"9_PARCLIP_ELAVL1MNASE_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"10_PARCLIP_ELAVL1A_hg19\", \"12_PARCLIP_EWSR1_hg19\", \"13_PARCLIP_FUS_hg19\", \"14_PARCLIP_FUS_mut_hg19\", \"15_PARCLIP_IGF2BP123_hg19\", \"16_ICLIP_hnRNPC_Hela_iCLIP_all_clusters\", \"17_ICLIP_HNRNPC_hg19\", \"18_ICLIP_hnRNPL_Hela_group_3975_all-hnRNPL-Hela-hg19_sum_G_hg19--ensembl59_from_2337-2339-741_bedGraph-cDNA-hits-in-genome\", \"19_ICLIP_hnRNPL_U266_group_3986_all-hnRNPL-U266-hg19_sum_G_hg19--ensembl59_from_2485_bedGraph-cDNA-hits-in-genome\", \"20_ICLIP_hnRNPlike_U266_group_4000_all-hnRNPLlike-U266-hg19_sum_G_hg19--ensembl59_from_2342-2486_bedGraph-cDNA-hits-in-genome\", \"21_PARCLIP_MOV10_Sievers_hg19\", \"22_ICLIP_NSUN2_293_group_4007_all-NSUN2-293-hg19_sum_G_hg19--ensembl59_from_3137-3202_bedGraph-cDNA-hits-in-genome\", \"23_PARCLIP_PUM2_hg19\", \"24_PARCLIP_QKI_hg19\", \"25_CLIPSEQ_SFRS1_hg19\",\"26_PARCLIP_TAF15_hg19\", \"27_ICLIP_TDP43_hg19\", \"28_ICLIP_TIA1_hg19\", \"29_ICLIP_TIAL1_hg19\", \"30_ICLIP_U2AF65_Hela_iCLIP_ctrl_all_clusters\", \"31_ICLIP_U2AF65_Hela_iCLIP_ctrl+kd_all_clusters\"]\n",
    "ionmf = [0.913, 0.714, 0.928, 0.925, 0.752, 0.939, 0.942, 0.954, 0.723, 0.974, 0.971, 0.893, 0.811, 0.955, 0.936, 0.942, 0.976, 0.826, 0.801, 0.796, 0.956, 0.868, 0.961, 0.877, 0.898, 0.910, 0.860, 0.943, 0.942, 0.915, 0.925]\n",
    "difference = np.subtract(score_list,ionmf)\n",
    "pd.set_option('display.max_columns', 4)\n",
    "\n",
    "\n",
    "              \n",
    "table_data = {'Score' : score_list, 'Difference':difference}\n",
    "table = pd.DataFrame(table_data, index = protein_list)\n",
    "\n",
    "print (table)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-4.13330110e-03,  2.70608034e-02, -7.23632202e-03,  8.68704100e-03,\n",
       "        2.27194680e-02, -2.51158423e-04,  2.22924000e-03,  3.43545150e-03,\n",
       "        3.73126986e-02,  4.17648938e-03, -1.03878750e-04,  5.44699277e-04,\n",
       "       -8.32807141e-03, -6.89828850e-03, -2.19468310e-05, -7.51118789e-03,\n",
       "       -1.71405011e-02,  1.86622449e-02,  5.90083686e-02,  3.16280316e-02,\n",
       "        5.03512313e-03,  3.86112899e-02, -1.05242674e-03, -1.18943413e-03,\n",
       "        2.88111594e-03, -1.27321143e-02,  1.24755976e-03, -1.23421215e-03,\n",
       "       -1.00392121e-02, -1.85024229e-02, -1.07080554e-02])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score_list1= [0.9166314660608545, 0.7355842739040614, 0.9156874064729882, 0.9110003605733115, 0.7106967639824682, 0.9175852831044603, 0.9532003478242389, 0.9609342519839978, 0.7175438955225013, 0.9260018057724553, 0.9309924438043105, 0.9323816311487496, 0.9071851247804773, 0.9831397242073813, 0.9433773325676775, 0.9346670168149185, 0.9593113189919608, 0.7925042050312222, 0.7068700113148783, 0.7681858296389412, 0.969496120005114, 0.812325268153552, 0.9779414360292946, 0.9469422043878664, 0.8703725801929352, 0.9702106147520017, 0.8781391345632541, 0.9387041550637198, 0.9376481211314986, 0.9335122789612825, 0.9341154218778888]\n",
    "score_list2 = [0.91249816495929, 0.7626450772807127, 0.9084510844575933, 0.9196874015687442, 0.7334162319501186, 0.9173341246812802, 0.9554295878289565, 0.9643697034835925, 0.7548565940975945, 0.9301782951531139, 0.9308885650545436, 0.932926330425737, 0.898857053374814, 0.976241435704717, 0.9433553857366352, 0.9271558289215133, 0.94217081786666, 0.81116644993974, 0.765878379881905, 0.7998138612249177, 0.9745312431338725, 0.8509365580449263, 0.976889009288503, 0.945752770259, 0.8732536961307217, 0.9574785004610682, 0.8793866943231377, 0.9374699429100511, 0.9276089090165114, 0.91500985610617, 0.9234073665024163]\n",
    "\n",
    "np.subtract(score_list2, score_list1)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
